src/entities/database.js

```diff
@@ -268,26 +305,170 @@ class DB implements Database {
+          inputs_address: inputAddresses,
+          inputs_amount: inputAmmounts,
+        }
+        : {}),
     }
     const now = new Date().toUTCString()
-    const query = Q.TX_INSERT.setFields(txDbFields)
-      .onConflict('hash', {
+
+    const onConflictArgs = []
+    if (upsert) {
+      onConflictArgs.push('hash', {
         block_num: blockNum,
         block_hash: blockHash,
         time: txUTCTime,
         tx_state: txStatus,
         last_update: now,
         tx_ordinal: tx.txOrdinal,
       })
+    }
+
+    const sql = Q.TX_INSERT.setFields(txDbFields)
+      .onConflict(...onConflictArgs)
       .toString()
-    this.#logger.debug('Insert TX:', query, inputAddresses, inputAmmounts)
-    await conn.query(query)
+    this.#logger.debug('Insert TX:', sql, inputAddresses, inputAmmounts)
+    await conn.query(sql)
     await this.storeTxAddresses(
       id,
       [...new Set([...inputAddresses, ...outputAddresses])],
     )
   }
 
+  async isTxExists(txId: string): boolean {
+    const sql = Q.sql.select().from('txs')
+      .field('1')
+      .where('hash = ?', txId)
+      .limit(1)
+      .toString()
+    const dbRes = await this.getConn().query(sql)
+    return dbRes.rows.length === 1
+  }
+
+  async isTxValid(txHash: string): Promise<boolean> {
+    const sql = Q.sql.select().from('txs')
+      .where('hash = ?', txHash)
+      .where('tx_state = ?', TX_STATUS.TX_SUCCESS_STATUS)
+      .toString()
+    const dbRes = await this.getConn().query(sql)
+    const [tx] = dbRes.rows
+    return Boolean(tx) && this.utxosForInputsExists(tx.inputs)
+  }
+
+  async queryPendingSet() {
+    const query = Q.sql.select().from(SNAPSHOTS_TABLE)
+      .field('tx_hash')
+      .where('block_height = ?', Q.sql.select().from(SNAPSHOTS_TABLE)
+        .field('MAX(block_height)'))
+      .where('status = ?', TX_STATUS.TX_PENDING_STATUS)
+      .union(Q.sql.select().from('txs')
+        .field('hash')
+        .where('tx_state = ?', TX_STATUS.TX_PENDING_STATUS)
+        .where('NOT EXISTS ?', Q.sql.select().from(SNAPSHOTS_TABLE)
+          .field('1')
+          .where('tx_hash = hash')))
+      .toString()
+    const dbRes = await this.getConn().query(query)
+    this.#logger.debug('queryPendingSet:', query, dbRes)
+    return _.map(dbRes.rows, 'tx_hash')
+  }
+
+  async groupPendingTxs(txs: Array<string>): Promise<[Array<string>, Array<string>]> {
+    const validTxs = []
+    const invalidTxs = []
+    for (const tx of txs) {
+      const isValid = await this.isTxValid(tx)
+      if (isValid) {
+        validTxs.push(tx)
+      } else {
+        this.#logger.info(`tx ${tx} inputs already spent`)
+        invalidTxs.push(tx)
+      }
+    }
+    return [validTxs, invalidTxs]
+  }
+
+  async groupPendingTxsForSnapshot(txHashes: Array<string>): Promise<[Array<string>, Array<string>]> {
+    const pendingSet = await this.queryPendingSet()
+    const txsInPendingState = pendingSet.filter(hash => !txHashes.includes(hash))
```

Comment:
There's a lodash function to filter arrays like this, I think it's also easier to read:
```
const txsInPendingState = _.difference(pendingSet, newConfirmedTxHashes)
```